File: Data Structures & Algorithms/foreign-dictionary/submission-1.cpp
class Solution {
public:
    unordered_map<char, unordered_set<char>> adj;
    unordered_map<char, bool> vis;
    string result;

    string foreignDictionary(vector<string>& words) {
        
        for(auto x: words){
            for(auto c: x){
                adj[c];
            }
        }

        for(int i = 1;i<words.size();i++){
            string w1 = words[i-1], w2 = words[i];
            int n = min(w1.size(), w2.size());
            
            if(w1.size() > w2.size() && w1.substr(0, n) == w2.substr(0, n)) 
                return "";
            
            for(int j=0;j<n;j++){
                if(w1[j]!=w2[j]){
                    adj[w1[j]].insert(w2[j]);
                    break;
                }
            }
        }
        
        for(auto x: adj){
            if(dfs(x.first)) 
                return "";
        }

        reverse(result.begin(), result.end());
        return result;
            
    }

    bool dfs(char c){
        
        if(vis.find(c) != vis.end()) return vis[c];

        vis[c] = 1;
        for(auto x: adj[c]){
            if(dfs(x)) return 1;
        }
        vis[c] = 0;
        result.push_back(c);
        return 0;

    }
};

File: Data Structures & Algorithms/regular-expression-matching/submission-1.java
class Solution {
    public boolean isMatch(String s, String p) {
        Boolean[][] dp = new Boolean[s.length() + 1][p.length() + 1];
        return dfs(0, 0, s, p, dp);
    }

    private boolean dfs(int i, int j, String s, String p, Boolean[][] dp) {
        int n = s.length(), m = p.length();
        
        if (j == m) {
            return i == n;
        }

        if (dp[i][j] != null) return dp[i][j];

        boolean currentMatch = (i < n) && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.');

        boolean result;

        if (j + 1 < m && p.charAt(j + 1) == '*') {
            result = dfs(i, j + 2, s, p, dp) || (currentMatch && dfs(i + 1, j, s, p, dp));
        } else {
            result = currentMatch && dfs(i + 1, j + 1, s, p, dp);
        }

        return dp[i][j] = result;
    }
}